consul/consul
# Store files with both Paperclip and ActiveStorage

## References

* [Paperclip has been deprecated for 3 years](https://thoughtbot.com/blog/closing-the-trombone)
* [Paperclip to Active Storage migration guide](https://github.com/thoughtbot/paperclip/blob/master/MIGRATING.md)
* [How to migrate to Active Storage without losing your mind](https://www.youtube.com/watch?v=tZ_WNUytO9o)
* This pull request implements the first step required to solve issue #2629 

## Objectives

* Make it possible to replace Paperclip with Active Storage in the future

## Notes

In order to migrate existing files from Paperclip to ActiveStorage, we need Paperclip to find out the files associated to existing database records. So we can't simply replace Paperclip with ActiveStorage.

That's why it's usually recommended to first run the migration and then replace Paperclip with ActiveStorage using two consecutive deployments.

However, in our case we can't rely on two consecutive deployments because we have to make an easy process so existing CONSUL installations don't run into any issues. We can't just release version 1.4.0 and 1.5.0 the same day and ask everyone to upgrade twice in a row.

Instead, we're following a different plan:

* We provide a Rake task (which will require Paperclip) to migrate existing files
* We still use Paperclip to generate link and image tags
* New files are handled using both Paperclip and ActiveStorage; that way, when we make the switch, we won't have to migrate them, and in the meantime they'll be accessible thanks to Paperclip
* After we make the switch (which will happen after releasing version 1.4.0), we'll update the `name` column in the active storage attachments tables in order to remove the `storage_` prefix

### Testing this pull request

Before merging this pull request, we need to test this scenario:

* Make sure we've got records with attachmets for all models handling attachments
* Run the `migrate_from_paperclip` task
* Attach files to both new and existing records from all models handling attachments
* Apply the changes in #4600 and run the `remove_paperclip_compatibility_in_existing_attachments` task
* Check images and documents attached before and after running the `migrate_from_paperclip` are correctly displayed

We need to do so with both a local storage and a remote storage.

## Release Notes

:warning: If you've added custom attachments, make sure to change these attachments so they use  with `has_attachment` instead of `has_attached_file`. Otherwise you might run into missing files when upgrading to version 1.5.0.

In version 1.5.0 we'll change the way we store attachments from Paperclip to Active Storage. In order to make the migration painless, version 1.4.0 stores new files using both systems. If you've changed anything related to the way attachments are handled, there's a chance you might run into some issues when executing `consul:execute_release_tasks` (which includes a task to copy existing files to Active Storage). In version 1.5.0 we'll drop Paperclip and use just Active Storage, so make sure to fix the issues or report them; otherwise you might run into missing files when upgrading to version 1.5.0.

After copying the existing files to Active Storage, **keep both the Active Storage files and the Paperclip files**. The Paperclip files are needed in version 1.4.0, and the Active Storage files will be needed in version 1.5.0.

If you're storing files using an external service (such as S3), make sure to edit the `config/storage.yml` file and configure the service. Sample configurations are provided for S3, GCS and Azure; uncomment them where appropriate and then edit your `config/secrets.yml` file with your credentials.
